File: hydra-haskell/src/main/haskell/Hydra/Dsl/Meta/Compute.hs
{-# LANGUAGE FlexibleContexts #-}

-- | Meta-DSL for constructing compute-related terms (Adapter, Coder, etc.)

module Hydra.Dsl.Meta.Compute where

import Hydra.Kernel
import Hydra.Dsl.AsTerm
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Meta.Core as Core

import qualified Data.Map as M
import qualified Data.Maybe as Y


adapter :: AsTerm c (Coder v1 v2) => TTerm Bool -> TTerm t1 -> TTerm t2 -> c -> TTerm (Adapter t1 t2 v1 v2)
adapter isLossy source target coderArg = record _Adapter [
  _Adapter_isLossy>>: isLossy,
  _Adapter_source>>: source,
  _Adapter_target>>: target,
  _Adapter_coder>>: asTerm coderArg]

adapterIsLossy :: TTerm (Adapter t1 t2 v1 v2) -> TTerm Bool
adapterIsLossy a = project _Adapter _Adapter_isLossy @@ a

adapterSource :: TTerm (Adapter t1 t2 v1 v2) -> TTerm t1
adapterSource a = project _Adapter _Adapter_source @@ a

adapterTarget :: TTerm (Adapter t1 t2 v1 v2) -> TTerm t2
adapterTarget a = project _Adapter _Adapter_target @@ a

adapterCoder :: TTerm (Adapter t1 t2 v1 v2) -> TTerm (Coder v1 v2)
adapterCoder a = project _Adapter _Adapter_coder @@ a

adapterWithCoder :: TTerm (Adapter t1 t2 v1 v2) -> TTerm (Coder v1 v2) -> TTerm (Adapter t1 t2 v1 v2)
adapterWithCoder a coder = adapter
  (Hydra.Dsl.Meta.Compute.adapterIsLossy a)
  (Hydra.Dsl.Meta.Compute.adapterSource a)
  (Hydra.Dsl.Meta.Compute.adapterTarget a)
  coder

adapterWithTarget :: TTerm (Adapter t1 t2 v1 v2) -> TTerm t2 -> TTerm (Adapter t1 t2 v1 v2)
adapterWithTarget a target = adapter
  (Hydra.Dsl.Meta.Compute.adapterIsLossy a)
  (Hydra.Dsl.Meta.Compute.adapterSource a)
  target
  (Hydra.Dsl.Meta.Compute.adapterCoder a)

coder :: TTerm (Context -> v1 -> Either (InContext Error) v2) -> TTerm (Context -> v2 -> Either (InContext Error) v1) -> TTerm (Coder v1 v2)
coder encode decode = record _Coder [
  _Coder_encode>>: encode,
  _Coder_decode>>: decode]

coderEncode :: TTerm (Coder v1 v2) -> TTerm (Context -> v1 -> Either (InContext Error) v2)
coderEncode c = project _Coder _Coder_encode @@ c

coderDecode :: TTerm (Coder v1 v2) -> TTerm (Context -> v2 -> Either (InContext Error) v1)
coderDecode c = project _Coder _Coder_decode @@ c

File: hydra-java/src/main/java/hydra/dsl/meta/Compute.java
package hydra.dsl.meta;

import hydra.util.Adapter;
import hydra.util.Coder;
import hydra.phantoms.TTerm;

import static hydra.dsl.meta.Phantoms.*;

/**
 * Meta-DSL for constructing computation-related terms (Adapter, Coder).
 *
 * <p>Mirrors Hydra.Dsl.Meta.Compute (Haskell).
 */
public interface Compute {

    // ============================================================
    // Adapter
    // ============================================================

    static Expr<Adapter> adapter(TTerm<?> isLossy, TTerm<?> source, TTerm<?> target, TTerm<?> coder) {
        return record(Adapter.TYPE_,
                field(Adapter.IS_LOSSY, isLossy),
                field(Adapter.SOURCE, source),
                field(Adapter.TARGET, target),
                field(Adapter.CODER, coder));
    }

    static Expr<Boolean> adapterIsLossy(TTerm<Adapter> a) {
        return apply(project(Adapter.TYPE_, Adapter.IS_LOSSY), a);
    }

    static Expr<?> adapterSource(TTerm<Adapter> a) {
        return apply(project(Adapter.TYPE_, Adapter.SOURCE), a);
    }

    static Expr<?> adapterTarget(TTerm<Adapter> a) {
        return apply(project(Adapter.TYPE_, Adapter.TARGET), a);
    }

    static Expr<?> adapterCoder(TTerm<Adapter> a) {
        return apply(project(Adapter.TYPE_, Adapter.CODER), a);
    }

    static Expr<Adapter> adapterWithCoder(TTerm<Adapter> a, TTerm<?> coder) {
        return adapter(adapterIsLossy(a), adapterSource(a), adapterTarget(a), coder);
    }

    static Expr<Adapter> adapterWithTarget(TTerm<Adapter> a, TTerm<?> target) {
        return adapter(adapterIsLossy(a), adapterSource(a), target, adapterCoder(a));
    }

    // ============================================================
    // Coder
    // ============================================================

    static Expr<Coder> coder(TTerm<?> encode, TTerm<?> decode) {
        return record(Coder.TYPE_,
                field(Coder.ENCODE, encode),
                field(Coder.DECODE, decode));
    }

    static Expr<?> coderEncode(TTerm<Coder> c) {
        return apply(project(Coder.TYPE_, Coder.ENCODE), c);
    }

    static Expr<?> coderDecode(TTerm<Coder> c) {
        return apply(project(Coder.TYPE_, Coder.DECODE), c);
    }
}
